### ISSUE TITLE
Cannot install npm packages with git+ssh

### ISSUE THREAD

**Issue by eamondawes** [ibmruntimes/node#99]
I am unable to install npm packages onto USS using the git+ssh syntax,  something I am able to do from Windows. This problem seems to be related to do with the function cb() never being called. In both of the below cases, I run the following commands in a new directory. I ensure I have ssh access to github from both environments. I am using [Rocket's version of git](http://www.rocketsoftware.com/product-categories/mainframe/git-for-zos), and  can provide a copy if need be for testing. I use expressjs as an example, but the behavior is the same for any package.

**Windows**

```
> node -v
v6.11.3
> npm --version
5.4.2
> git --version
git version 2.14.1.windows.1

> ssh -T git@github.com
Hi eamondawes! You've successfully authenticated, but GitHub does not provide shell access.

> npm init
**Accepting default values, output omitted for brevity**
About to write to C:\tmp\package.json:

{
  "name": "tmp",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC"
}

Is this ok? (yes)
> npm install --save-dev git+ssh://git@github.com:expressjs/express.git
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN tmp@1.0.0 No description
npm WARN tmp@1.0.0 No repository field.

+ express@4.16.2
added 48 packages in 3.03s
```

**USS**

```
> node --version
v6.11.2
> npm --version
5.4.2
> git --version
git version 2.3.5_zos_b015

> ssh -T git@github.com
Hi eamondawes! You've successfully authenticated, but GitHub does not provide shell access.

> npm init
**Accepting default values, output omitted for brevity**
About to write to /u/ts5558/tmp/package.json:

{
  "name": "tmp",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC"
}

Is this ok? (yes) yes
> npm install --save-dev git+ssh://git@github.com:expressjs/express.git
Unhandled rejection Error: Command failed: /rsusr/ported/bin/git checkout ������Ù���/���/���Ù��/��Ô��������conv-lite-0.4.19.tgz extracted to /u/ts5558/tmp/
���?ʚ��/�����Ā▒▒�]��]��]��]��]��]Cr�����]��]��]���]��]��]��]��]Cr�����]��]��]Cm��]��]��]��]��]��]��]��]▒▒����>?Ȁ_/��ǀ/>`���%��ˉ�,>?�>��?�����
    at ChildProcess.exithandler (child_process.js:198:12)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:902:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)

npm ERR! cb() never called!

npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! A complete log of this run can be found in:
npm ERR!     /u/ts5558/.npm/_logs/2017-12-15T15_43_09_069Z-debug.log
```

**Some context with silly debugging for npm install**

```
npm sill pacote mime-db@https://registry.npmjs.org/mime-db/-/mime-db-1.30.0.tgz extracted to /u/ts5558/tmp/node_modules/.staging/mime-db-8cc4ec09 by content address 402ms
npm sill pacote ipaddr.js@https://registry.npmjs.org/ipaddr.js/-/ipaddr.js-1.5.2.tgz extracted to /u/ts5558/tmp/node_modules/.staging/ipaddr.js-13851bd8 by content address 406ms
npm sill pacote debug@https://registry.npmjs.org/debug/-/debug-2.6.9.tgz extracted to /u/ts5558/tmp/node_modules/.staging/debug-4972d4a7 by content address 402ms
npm sill pacote qs@https://registry.npmjs.org/qs/-/qs-6.5.1.tgz extracted to /u/ts5558/tmp/node_modules/.staging/qs-9cf97c2b by content address 399ms
npm sill pacote iconv-lite@https://registry.npmjs.org/iconv-lite/-/iconv-lite-0.4.19.tgz extracted to /u/ts5558/tmp/node_modules/.staging/iconv-lite-7e3b2d0a by content address 418ms
Unhandled rejection Error: Command failed: /rsusr/ported/bin/git checkout ������Ù���/���/���Ù��/��Ô��������conv-lite-0.4.19.tgz extracted to /u/ts5558/tmp/
���?ʚ��/�����Ā▒▒�]��]��]��]��]��]Cr�����]��]��]���]��]��]��]��]Cr�����]��]��]Cm��]��]��]��]��]��]��]��]▒▒����>?Ȁ_/��ǀ/>`���%��ˉ�,>?�>��?�����
    at ChildProcess.exithandler (child_process.js:198:12)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:902:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)

npm ERR! cb() never called!

npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! A complete log of this run can be found in:
npm ERR!     /u/ts5558/.npm/_logs/2017-12-15T15_59_44_388Z-debug.log
```

**Comment by mmallick-ca** [comment]
As far as I know Rocket's port of git is built with -qascii and as long as the appropriate environment variables are set the environment will convert all characters from ascii to ebcdic and vice versa. However there is an exception  when it comes to sockets and I believe the child processes in Node redirect their stdout to sockets which the parent consumes. 

There is a workaround which will help you get past this. 

I suggest you wrap your git executable with a shell script and name it 'git'. 
> cat /home/mallick/bin/git
#!/bin/sh
<ABSOLUTE_PATH_TO_YOUR_GIT_BINARY>/git $@ | iconv -f ISO8859-1 -t IBM-1047 | cat 

The script  will pipe the output of the original git executable through the 'iconv' utility and then pass it through cat .Make sure to place this shell script in your PATH such that the script is executed when git is invoked.

> echo $PATH
**/home/mallick/bin:**/home/mallick/nodejsGAcompiler/bin:/home/mallick/node-v6.11.2-os390-s390x/bin:/home/barboza/python-2.7.13/python-2017-04-12-py27/python27/bin:/bin

Verify that if git is invoked it will invoke this wrapper 
> which git
/home/mallick/bin/git

Now try 
>npm install --save-dev git+ssh://git@github.com:expressjs/express.git
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN issue@1.0.0 No description
npm WARN issue@1.0.0 No repository field.

+ express@4.16.2
added 48 packages in 4.349s

Hope this helps.

**Comment by eamondawes** [comment]
This works, I am able to download packages using git+ssh, but other commands (git log, git status e.g.) are now displayed as garbage. Having to switch between two different **git** commands is a hassle. Will this problem will have a more permanent solution in the future?

**Comment by mmallick-ca** [comment]
Do you have _BPXK_AUTOCVT set to "ON"?  If you do and you are using the standard shell in USS then the regular git commands should not output garbage even when you are using the wrapper.
There is also the option of creating an alias for the non_wrapped git.

Rockets port of git is as an ASCII application so you will need the aforementioned environment variable set and also add FILETAG(AUTOCVT,AUTOTAG) to your _CEE_RUNOPTS environment variable.

As far as a permanent solution is concerned, I just want to clarify that this is not a defect or an issue in the Node runtime per say. Its more a consequence of the fact that the USS does not autoconvert content piped from an ascii application to an ebcdic application when fed through sockets. We will need to discuss with the z/OS / USS team to understand why this limitation exists and whether it can be resolved.

**Comment by rich335z** [comment]
I have this issue, as well. I encountered this with a custom node module (developed for internal use). I worked around the issue by (prematurely) publishing my module to our local npm registry and installing it as a "normal" npm dependency, but this is not an acceptable approach. I would expect to be able to install this kind of module from its git repo until it's been vetted and deemed acceptable for publishing.

Having said that... I created the wrapper described above, and there are still issues:

Firstly, I can not validate my wrapper script with the `which` command under the default USS shell nor `bash`, as that command doesn't exist for those shells. As far as I can tell, the `which` command is only available with the t-shell. However, the `type` command is available with any of these shells, so that's probably a better option to recommend.

Also, after setting up the wrapper script, my _general_ git commands now produce garbage:
```
rjack11:/u/rjack11/richapp2/richapp2: >git log
??????@???????????????????????????????????????z@ى??@с?????@L???????|??????????K???`????K???nā??z@@@㤅@ƅ?@?@??z??z??@????@`????@@@@م???@??????a???????????@?????????􁄄?󆅆?????????????????????????z@ى??@с?????@L???????|??????????K???`????K???nā??z@@@Ԗ?@ƅ?@?@??z??z??@????@`????@@@@م???@??????@???@?`?????????????@???????????????????????????????????????????z@ى??@с?????@L???????|??????????K???`????K???nā??z@@@Ԗ?@ƅ?@?@??z??z??@????@`????@@@@ɕ?????@Ö????rjack11:/u/rjack11/richapp2/richapp2: >git status
֕@??????@??????薤?@??????@??@??`??`????@????@}??????a??????}K???????@??@??????k@???????@?????????@?????rjack11:/u/rjack11/richapp2/richapp2: >
```
And I _do_ have the recommended settings in place:
```
_BPXK_AUTOCVT=ON
_CEE_RUNOPTS="FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)"
```

Now, my `git+ssh` command encounters a different issue, but doesn't produce garbage anymore:
```
rjack11:/u/rjack11/test/nodetest: >npm install --save git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git
npm ERR! setPermissions Failed to change git repository ownership under npm cache for /u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144
npm ERR! OS/390 25.00
npm ERR! argv "/u/rjack11/node-v6.11.2-os390-s390x/bin/node" "/u/rjack11/node-v6.11.2-os390-s390x/bin/npm" "install" "--save" "git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git"
npm ERR! node v6.11.2
npm ERR! npm  v3.10.10
npm ERR! path /u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144
npm ERR! code ENOENT
npm ERR! errno -129
npm ERR! syscall scandir

npm ERR! enoent ENOENT: no such file or directory, scandir '/u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144'
npm ERR! enoent ENOENT: no such file or directory, scandir '/u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! Please include the following file with any support request:
npm ERR!     /u/rjack11/test/nodetest/npm-debug.log
rjack11:/u/rjack11/test/nodetest: >cat /u/rjack11/test/nodetest/npm-debug.log
0 info it worked if it ends with ok
1 verbose cli [ '/u/rjack11/node-v6.11.2-os390-s390x/bin/node',
1 verbose cli   '/u/rjack11/node-v6.11.2-os390-s390x/bin/npm',
1 verbose cli   'install',
1 verbose cli   '--save',
1 verbose cli   'git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git' ]
2 info using npm@3.10.10
3 info using node@v6.11.2
4 silly loadCurrentTree Starting
5 silly install loadCurrentTree
6 silly install readLocalPackageData
7 silly fetchPackageMetaData git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git
8 silly fetchOtherPackageData git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git
9 silly cache add args [ 'git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git', null ]
10 verbose cache add spec git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git
11 silly cache add parsed spec Result {
11 silly cache add   raw: 'git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git',
11 silly cache add   scope: null,
11 silly cache add   escapedName: null,
11 silly cache add   name: null,
11 silly cache add   rawSpec: 'git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git',
11 silly cache add   spec: 'ssh://gecgithub01.walmart.com:zPAC/zuid-node.git',
11 silly cache add   type: 'git' }
12 verbose addRemoteGit caching git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git
13 silly tryClone cloning git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git via git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git
14 verbose tryClone git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144 not in flight; caching
15 verbose correctMkdir /u/rjack11/.npm/_git-remotes correctMkdir not in flight; initializing
16 info git [ 'clone',
16 info git   '--template=/u/rjack11/.npm/_git-remotes/_templates',
16 info git   '--mirror',
16 info git   'gecgithub01.walmart.com:zPAC/zuid-node.git',
16 info git   '/u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144' ]
17 verbose mirrorRemote git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git git clone gecgithub01.walmart.com:zPAC/zuid-node.git
18 verbose correctMkdir /u/rjack11/.npm/_git-remotes correctMkdir not in flight; initializing
19 error setPermissions Failed to change git repository ownership under npm cache for /u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144
20 silly fetchPackageMetaData Error: ENOENT: no such file or directory, scandir '/u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144'
20 silly fetchPackageMetaData     at Error (native)
20 silly fetchPackageMetaData  error for git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git { Error: ENOENT: no such file or directory, scandir '/u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144'
20 silly fetchPackageMetaData     at Error (native)
20 silly fetchPackageMetaData   errno: -129,
20 silly fetchPackageMetaData   code: 'ENOENT',
20 silly fetchPackageMetaData   syscall: 'scandir',
20 silly fetchPackageMetaData   path: '/u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144' }
21 silly rollbackFailedOptional Starting
22 silly rollbackFailedOptional Finishing
23 silly runTopLevelLifecycles Finishing
24 silly install printInstalled
25 verbose stack Error: ENOENT: no such file or directory, scandir '/u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144'
25 verbose stack     at Error (native)
26 verbose cwd /u/rjack11/test/nodetest
27 error OS/390 25.00
28 error argv "/u/rjack11/node-v6.11.2-os390-s390x/bin/node" "/u/rjack11/node-v6.11.2-os390-s390x/bin/npm" "install" "--save" "git+ssh://gecgithub01.walmart.com:zPAC/zuid-node.git"
29 error node v6.11.2
30 error npm  v3.10.10
31 error path /u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144
32 error code ENOENT
33 error errno -129
34 error syscall scandir
35 error enoent ENOENT: no such file or directory, scandir '/u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144'
36 error enoent ENOENT: no such file or directory, scandir '/u/rjack11/.npm/_git-remotes/git-ssh-gecgithub01-walmart-com-zPAC-zuid-node-git-7418b144'
36 error enoent This is most likely not a problem with npm itself
36 error enoent and is related to npm not being able to find a file.
37 verbose exit [ -129, true ]
```

Regardless of my ultimately particular `npm install git+ssh` issue here, I would expect general `npm install` behavior to be the same as elsewhere. A work-around is OK during beta testing, but not when it breaks other basic functionality. This is obviously a z/OS-specific problem with the node port, and should be addressed accordingly if we want this to be a viable product.

**Comment by mmallick-ca** [comment]
I am in complete agreement with both of you with regards to the fact that the end user experience and behavior should be seamless. What I am trying to say though is that there are 4 pieces of technologies interacting here. NPM, Node, Git and the z/OS Unix System Services. Generally speaking, when the appropriate environment variables are set, any output from Git converted to EBCDIC when piped to a non-ASCII application. Unfortunately 
there is an exception when it comes to Unix sockets. Node's implementation uses these sockets to pipe output from its child process instances to its parent processes. 

The irony here is that in order to enable seamless behavior what we do in Node  is convert any output we receive from these sockets from EBCDIC->ASCII. This is based on the assumption that since we are running on an EBCDIC based system, programs  compiled to run on the system will produce EBCDIC output. However the exception here of course is anything compiled with -qascii which is how Rocket has built its port of Git. 

Ultimately if USS had a facility for us to detect the type of application(i.e.ASCII based or EBCDIC based) with a tag, similar to what it does for files, we could make a more informed decision on a case by case basis. Better yet if USS did not make an exception to auto-conversion when it came to piping output via sockets.  This is why I personally feel that the right technological solution is in the USS layer. We are reaching out to our USS colleagues to get their insight into this issue

**Comment by zorts** [comment]
This issue is only going to get worse over time, as more of the Rocket ports transition to using -qascii.

It seems odd to me that node uses sockets, as opposed to pipes (which support autoconversion), to exchange data with child processes. I would think that would be easy to change.

**Comment by mmallick-ca** [comment]
Yes, that is another option, to change the IPC mechanism from sockets to pipes. We need to look into the current implementation to see how big a change this is. However the risk we run there is that we diverge from the community implementation.

**Comment by jBarz** [comment]
Pipes are unidirectional so there is additional overhead per child process with tracking read and write using 2 file descriptors instead of the one socket. I don't expect this change to be trivial.

**Comment by ghost** [comment]
Any new ideas on how to solve this problem?  Do you have an estimate on the proposed switch over from sockets to pipes?  The proposed workaround of aliasing Git will not work.  Our build script could be coded to use an alias, but an individual developer's workflow would still be affected negatively.  We must support both scenarios from day 1.

**Comment by mmallick-ca** [comment]
We investigated the pipes-> sockets switch and it seems there are other technical reasons why sockets are used to communicate with child processes in Nodes(passing sessions from parent -> child processes, passing ancillary information about filedescriptors etc). We have communicated with our USS colleagues and they are investigating solutions but I suspect any solution will not be available in the short term.

Another potential work-around that we could implement in the Node port is to detect that a child process is executing the Git application and just do the appropriate conversions before passing it onto the socket. This would work for this specific issue but will not work other -qascii compiled programs. So I guess the question is are there other Rocket applications(that have been built with -qascii) that you might be using with Node?

**Comment by zorts** [comment]
Everything Rocket ports to USS from this point on (and for about the last 18 months) is built -qascii.

**Comment by jBarz** [comment]
In trying to investigate this, I tried to run the following git commands that emulate what npm is doing.

```
git clone  '--mirror' 'ssh://git@github.com/expressjs/express.git'
git clone express.git newdir
```

I get an error msg.
```
Cloning into '/home/barboza/node.v6/temp/rocket/new'...
error: cannot run git-upload-pack '/home/barboza/node.v6/temp/rocket/express.git/': EDC5129I No such file or directory.
fatal: unable to fork
```

However, i also noticed that my git version is different from yours.
git version 2.3.5_os390_b011

Do you know if there is a way to work around this?

**Comment by zorts** [comment]
I was able to successfully run the commands above, but, as you noted, with the more recent version of git:

```
bash-4.3$ git clone '--mirror' 'ssh://git@github.com/expressjs/express.git'
Cloning into bare repository 'express.git'...
bash-4.3$ git clone express.git newdir
Cloning into 'newdir'...
done.
bash-4.3$ type git
git is hashed (/rsusr/ported/bin/git)
bash-4.3$ type git-upload-pack
git-upload-pack is /rsusr/ported/bin/git-upload-pack
bash-4.3$ git --version
git version 2.3.5_zos_b015
```
The important parts of my environment look like this:

```
LANG=C
LIBPATH=/usr/lib:/rsusr/ported/lib:/rsusr/local/lib:/usr/lpp/Printsrv/lib
PATH=/u/tsjlc/bin:/rsusr/ported/bin:/rsusr/local/bin:/bin:/usr/sbin:/data/mvscvs/bin:/usr/lpp/Print\
srv/bin:/rsusr/java/IBM/J6.0/bin
SHELL=/rsusr/ported/bin/bash
STEPLIB=none
TZ=EST5EDT4
_BPXK_AUTOCVT=ON
_BPXK_JOBLOG=STDERR
_CEE_RUNOPTS=FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)
_EDC_ADD_ERRNO2=1
```

**Comment by ghost** [comment]
>> So I guess the question is are there other Rocket applications(that have been built with -qascii) that you might be using with Node?

Git is the only requirement at this point.  I bet Python will be needed, as it's used by the gyp build tool.

**Comment by jBarz** [comment]
fwiw, I made some changes to the original in-built npm (3.10.10) that allowed me to do an `npm install git+ssh//...`
https://github.com/jBarz/node/commit/f28a6882b8d6276f2b374cb7ab106cb9bcac53dc

You could try and copy these files to your local installation and see if works for you.
[add-remote-git.js](https://raw.githubusercontent.com/jBarz/node/f28a6882b8d6276f2b374cb7ab106cb9bcac53dc/deps/npm/lib/cache/add-remote-git.js) -> lib/node_modules/npm/lib/cache/add-remote-git.js
[git.js](https://raw.githubusercontent.com/jBarz/node/f28a6882b8d6276f2b374cb7ab106cb9bcac53dc/deps/npm/lib/utils/git.js) -> lib/node_modules/npm/lib/utils/git.js
[configure.js](https://github.com/jBarz/node/blob/f28a6882b8d6276f2b374cb7ab106cb9bcac53dc/deps/npm/node_modules/node-gyp/lib/configure.js) -> lib/node_modules/npm/node_modules/node-gyp/lib/configure.js

Unfortunately, this means you have to use the npm that comes with node.
We are continuing to investigate a solution that will work with any npm version.

**Comment by jBarz** [comment]
This issue has been resolved since v 6.12.3.2
